prow/pjutil/filter.go

```diff
@@ -97,7 +97,8 @@ func FilterPresubmits(filter Filter, changes config.ChangedFilesProvider, branch
 // RetestFilter builds a filter for `/retest`
 func RetestFilter(failedContexts, allContexts sets.String) Filter {
 	return func(p config.Presubmit) (bool, bool, bool) {
-		return failedContexts.Has(p.Context) || (!p.NeedsExplicitTrigger() && !allContexts.Has(p.Context)), false, true
+		failed := failedContexts.Has(p.Context)
+		return failed || (!p.NeedsExplicitTrigger() && !allContexts.Has(p.Context)), false, failed
```

Comment:
> I still had to change even the `RetestFilter` for `default` to be `true` only for the failing jobs, not for the ones that get triggered by the "would run automatically but didn't yet" part of the `RetestFilter` condition.

I don't understand why this would be needed.

Reply:
This is needed because `/retest` triggers two kinds of jobs ([doc](https://github.com/kubernetes/test-infra/blob/master/prow/jobs.md#triggering-jobs-with-comments), [comment](https://github.com/kubernetes/test-infra/blob/4afd9ba744d65d6a3ff028118a4adb1a00026e7d/prow/plugins/trigger/generic-comment.go#L171-L174)):

1. The ones that executed in the past and posted a failing status. We **do** want to ignore usual trigger conditions b/c we want to get rid of the failing status.
2. The ones that did not execute in the past (didn't post a status) but would be now triggered b/c of the trigger conditions. This part is basically `/test all` but only for jobs that did not run yet. We *don't* want to ignore trigger conditions for these, otherwise we would trigger **all** jobs.

I don't think the second case actually happens that often -- the case should be preempted by `statusreconciler`, but I guess it can apply for some corner cases after GH outages where statuses are not reported correctly and the like.

Reply:
Ok, I see how this makes sense now when paired with the change to `Presubmit.ShouldRun`.